File: Foundations of Programming/15-interpreterPython/Problema11.py

# Write a simple interpreter which understands "+", "-", and "*" 
# operations. Apply the operations in order using command/arg 
# pairs starting with the initial value of `value`. 
# If you encounter an unknown command, return -1.

# interpret(1, ['+'], [1]) → 2
# interpret(4, ['-'], [2]) → 2
# interpret(1, ['+', '*'], [1, 3]) → 6

def interpret(value, commands, args):
    if len(commands) != len(args):
        return -1
    
    result = value

    for i in range(len(commands)):
        if commands[i] != "+" and commands[i] != "-" and commands[i] != "*":
           return -1

        if commands[i] == "+":
            result = result + args[i]
        elif commands[i] == "-":
            result = result - args[i]
        elif commands[i] == "*":
            result = result * args[i]
        else:
            return -1

    return result

# print(interpret(1, ['+'], [1]))
# print(interpret(4, ['-'], [2]))
print(interpret(1, ['+', '*'], [1, 3]))

File: Foundations of Programming/14-interpreterJava/Problema11.java
/*
Write a simple interpreter which understands "+", "-", and "*" 
operations. Apply the operations in order using command/arg 
pairs starting with the initial value of `value`. 
If you encounter an unknown command, return -1.

interpret(1, ["+"], [1]) → 2
interpret(4, ["-"], [2]) → 2
interpret(1, ["+", "*"], [1, 3]) → 6
*/
public class Problema11 {
    public static int interpret(int value, String[] commands, int[] args) {         
        if (commands.length != args.length)
            return -1;

        int result = value;
        
        for (int i=0; i< commands.length; i++) {
            if (!commands[i].equals("+") && 
                !commands[i].equals("-") &&
                !commands[i].equals("*"))
                return -1;
            
            switch(commands[i]) 
            { 
                case "+": 
                    result = result + args[i];
                    break; 
                case "-": 
                    result = result - args[i]; 
                    break; 
                case "*": 
                    result = result * args[i];
                    break; 
                default: 
                    return -1; 
            }
        }
        
        return result;
    }

    public static void main(String[] args){       
        // int value = 1;
        // String[] commands = {"+"};
        // int[] arg = {1};

        // int value = 4;
        // String[] commands = {"-"};
        // int[] arg = {2};

        int value = 1;
        String[] commands = {"+", "*"};
        int[] arg = {1, 3};

		int res = interpret(value, commands, arg);
		
		System.out.println(res);
	}
}